### ISSUE TITLE
Downstream component error being swallowed by react-apollo

### ISSUE THREAD

**Issue by cannoneyed** [apollostack/react-apollo#326]
## Steps to Reproduce

This is a bit of a head scratcher, but it seems like in certain cases errors in the render methods of child components are being swallowed by react-apollo, showing up in the `error.networkError`property of the `data` object. Here's the best approximation of code to reproduce:

```javascript
const ChildComponent = ({ user }) => {
    return (
        <div>{ user.property.doesntExist }</div>
    )
}

const Query = gql`
query UserQuery () {
    user {
        fullName
    }
}
`

@graphQl(Query)
class SmartComponent extends Component  {
    render() {
        const { loading, user } = this.props.data
        if (loading) {
            return <Loading />
        } else {
            return <ChildComponent user={user}
        }
    }
}
```

Unfortunately, the type error of accessing a property on undefined (`user.property.doesntExist`) in the child property is swallowed by apollo. You can find the error by logging `this.props.data` in the parent component, and you'll see this:

![screen shot 2016-11-14 at 1 44 54 pm](https://cloud.githubusercontent.com/assets/7818584/20284183/ed953c4c-aa70-11e6-9c65-e2c31c991b73.png)

### Expected Behavior

I'm not 100% sure if this is happening specifically in my use case or if there's something more general here. I think it's 100% not correct for an error in a child component to wind up being handled by Apollo's network error handling, and would expect that an error thrown by downstream (child) components would be thrown as usual 

### Version
- apollo-client@0.5.0
- react-apollo@0.5.16

**Comment by cannoneyed** [comment]
Closing because it's fixed in 0.6.0! :D

**Comment by cannoneyed** [comment]
OK, So it turns out that 0.6.0 doesn't fix the issue, it just silences the issue (arguably much worse. The same code as listed above lead to a much more inscrutable downstream error being thrown:

![screen shot 2016-11-14 at 3 18 58 pm](https://cloud.githubusercontent.com/assets/7818584/20286842/b20e5ade-aa7d-11e6-9de1-cfc5d151e920.png)

Seems like there's some issues with exception handling in the logic telling the child components to update

**Comment by cannoneyed** [comment]
In `graphql.tsx` ([here](https://github.com/apollostack/react-apollo/blob/master/src/graphql.tsx#L389-L396)), it seems the following thing that is warned about in a comment is actually happening...

```javascript
        ...
        /*
          Since `setState()` can throw an error if the child had a render error,
          we can't use the `error` part of the query subscription. If we do, all children
          errors are captured as NetworkErrors which isn't true / helpful.
          Instead, we subscribe to the store for network errors and re-render that way
        */
        this.querySubscription = this.queryObservable.subscribe({ next, error: handleError });
      }
```

**Comment by cannoneyed** [comment]
OK. Think I have something of a handle on what's happening here.

In `apollo-client/src/core/QueryManager.tsx` (code truncated)
```typescript
public queryListenerForObserver(
  ...
  if (observer.next) {
    ...
    try {
      observer.next(this.transformResult(resultFromStore))
  }
```

When the next result (`this.transformResult(resultFromStore)`) is returned from the store, it's passed as an argument into the observer's `next` method, which is defined below in `react-apollo`

In `react-apollo/src/graphql.tsx` (code truncated)
```typescript
subscribeToQuery(props): boolean {
  ...
  const next = (results: any) => {
    ...
    this.forceRenderChildren()
```

The `next` method provided to the `ObservableQuery` observer is invoked when the result of the query is received. This triggers an invocation of `this.forceRenderChildren`, which then leads to the `typeError` when the child component is rendered. This error is handled by the first `queryListenerForObserver` method, which calls `observer.next` in a `try/catch` block:

In `apollo-client/src/core/QueryManager.tsx` (code truncated)
```typescript
public queryListenerForObserver(
  ...
  if (observer.next) {
    ...
    try {
      observer.next(this.transformResult(resultFromStore))
    } catch (error) {
      observer.error(new ApolloError({
        networkError: error,
      }));
```

The `error` method declared in `react-apollo` has conditional logic for handling errors of type`ApolloError`, but any error anywhere in the render chain is getting cast to an ApolloError, which means that Apollo is swallowing them

In `react-apollo/src/graphql.tsx` (code truncated)
```typescript
const handleError = (error) => {
  if (error instanceof ApolloError) return next({ error });
  throw error;
};
``` 

I'm not exactly sure what to do about this, but I'm going to keep poking around at the internals to learn more about what the intended behavior here is. Anyway, having child component render errors get cast to an `ApolloError` (and a `networkError` property at that!) is kind of a deal breaker, so I'd really like to figure this one out.

**Comment by cannoneyed** [comment]
This is an issue with `apollo-client`. I've submitted a PR resolving it

https://github.com/apollostack/apollo-client/pull/910

**Comment by cannoneyed** [comment]
Resolved in apollostack/apollo-client#910. Closing
